### PR TITLE
Carburettor icing and other engine related improvements

### PR DESCRIPTION
**Pull request by gilbertohasnofb**
The carburetor icing system has been reviewed and several things have been improved:

- adjusted ice accumulation rate in icing conditions
- no ice accumulation when the engine is not running
- adjusted effectiveness of carburetor heat, which now is related to the engine exhaust temperature
- engine temperature affects ice formation (the warmer the engine, the lower the icing rate)
- the engine now coughs when the ice is melting, and the amount of coughing is proportional to the quantity of ice (i.e. severe icing accumulation == severe coughing)
- leaning when engine is coughing due to ice melting can help with coughing
- coughing can also happened due to fuel contamination with water, and the coughing interval is also proportional to the quantity of water
- lowered limit of critical oil level (when engine will quit) to below 2 quarts, according to Lycoming's manual
- critical oil will not cause the engine to cough, it will simply stop

Full discussion here: https://github.com/c172p-team/c172p-detailed/pull/826

### CODE REVIEW

**Comment by gilbertohasnofb** [comment]
@wlbragg when this gets merged, would you post these bullet points in the forum?

**Comment by gilbertohasnofb** [comment]
I have also updated the wiki page with some more info about the carb icing and other things (just two phrases here and there)

**Comment by wlbragg** [comment]
I've reviewed this and tested it. I'm OK to merge it. The only comment I have to make is the same one that @onox made on one of my last PR's that involved the use of a timer. He suggested that I adapt the "UpdateLoop" instead of a raw settimer. In fact the issue has yet to be deleted even though the PR was merged.
>This code should use the UpdateLoop instead of the raw settimer() calls. See TiedownPositionUpdater in Nasal/tiedowns.nas for an example.

See :https://github.com/c172p-team/c172p-detailed/issues/811
Maybe we can close two for one here with a clarification from *onox. I hate to leave these issues open, yet merged, because of something like this.

**Comment by gilbertohasnofb** [comment]
@wlbragg thanks for the feedback. So if all is good except some detail in the code then that's already moving forward :smile: 

But concerning the `settimer` thing, I believe the situation here is different than in that other issue you referenced: there you are using `settimer` to update the function, here I am simply having a small fixed 0.25 seconds delay between killing and un-killing the engine inside the cough function. Then notice that for the `coughing_timer` (which gives the interval between consecutive coughs) I am using a `maketimer` to control the update interval (in the `coughing_timer` function) which is then applied to the `engine_coughing` function, based on the fourth example from http://wiki.flightgear.org/Nasal_library#maketimer.28.29

So I think what I did looks fine, but I would be happy if @onox or anyone else would let me know if there is any problem.

**Comment by wlbragg** [comment]
All is good then as far as i am concerned.

>You're not authorized to merge this pull request. 

Do you have any idea as to why I am not authorized to merge anymore? Did I miss a staff meeting again? :laughing:

**Comment by gilbertohasnofb** [comment]
> why I am not authorized to merge anymore? Did I miss a staff meeting again?

@wlbragg that's quite bizarre! There was no meeting since we last created the c172p-team GitHub organization. I had a look at it and you do have write access, so I am puzzled why you can't merge this PR. I don't want to screw anything up and so since I am unfamiliar with these organizations, let's ask @onox to take a look at it, what do you think?

**Comment by wlbragg** [comment]
Yeah, it'll have to wait for @onox. I logged in and out, no change.

**Comment by onox** [comment]
@gilbertohasnofb I see some magic numbers changed. Where do these come from? In the bullet list in the first post you can get a sense of how the carburetor sorta works. Perhaps you can document the code a bit with the text from the first post. Imagine Github going down or the project moves to another service or you're working offline.

Using `maketimer` is good because you decouple the timer management from the actual loop code. `settimer` is kinda fire-and-forget. With `maketimer` you can restart timers for example, which you are doing in the code. Generally, for big systems that run continuously you could use UpdateLoop (it uses `maketimer` internally) because it takes care of bunch of signals (fdm-initialized, reinit). It also calls the `update` function of your component with the elapsed time (dt) as an argument. Might not be needed here.

**Comment by onox** [comment]
@wlbragg Try again if you can merge the PR.

**Comment by onox** [comment]
(after @gilbertohasnofb has added some extra comments if he wish)

**Comment by onox** [comment]
@wlbragg `master` is a protected branch and in addition I had enabled "Restrict who can push this branch". The reason was we always use PR's and so that some contributors wouldn't have to worry about accidentally pushing a messy graph/tree of commits. (Which would require interactive rebasing or a new feature branch when @onox spots such things 😉) We can always uncheck that checkbox at the bottom of course.

c172p-master

**Comment by gilbertohasnofb** [comment]
> I see some magic numbers changed. Where do these come from? In the bullet list in the first post you can get a sense of how the carburetor sorta works. Perhaps you can document the code a bit with the text from the first post. Imagine Github going down or the project moves to another service or you're working offline.

I tried talking about what I did in the issues and PRs involving this system, as well as adding quite a lot of comments in the code, but these are not centralized in a single place and indeed might be hard to follow. How would you suggest I document these?

As for the magic numbers, keep in mind that we don't have any data with measurement of ice and its effect in the carburettor, so this is all guess work. E.g. the loss os power due to icing is "guessed" until it felt right, same for the accumulation and efficacy of carb heat. What comes from data is only which situations icing occur, but how much icing is also guessed (and as we found out, different engines and different planes react differently to all this, and there is really no source I could find to really quantify those). So basically I just made sure that in severe icing you will experience carb icing in very few minutes, and I made sure that carb heat is still effective. From there it's all about testing and adjusting, which is what I did together with @kentrosaur and @wlbragg. So let me know about the documentation and I can work on it when I have time.

> Might not be needed here.

:+1: Thanks for the clarification about `maketimer` and `settimer`, and I am glad to hear it seems to be fine here

**Comment by onox** [comment]
I meant things like "engine temperature affects ice formation (the warmer the engine, the lower the icing rate)" and "lowered limit of critical oil level (when engine will quit) to below 2 quarts, according to Lycoming's manual".

**Comment by gilbertohasnofb** [comment]
>  "engine temperature affects ice formation (the warmer the engine, the lower the icing rate)"

That one is also use the _guess'n test_ method. I simply added a small factor given by the oil temperature (which is one way to gauge engine temp), and this factor decreases the icing rate a bit.

> "lowered limit of critical oil level (when engine will quit) to below 2 quarts, according to Lycoming's manual"

This one is quite simple. We had a critical value of just below 5 quarts of oil that when reached would kill the engine. This was given by the conservative figure that Cessna uses in their POHs, but if you read it carefully they are simply not recommending you to take off with less than 5 quarts (meaning 5 is still fine for short flights). So then I went to read Lycoming's manual for our engine, and they can actually guarantee that the engine is fine when the oil level in the sump is at least 2 quarts, so I lowered our "kill value" to 1.97. Link to Lycoming's manual and page of the source can be found in the discussion here: https://github.com/c172p-team/c172p-detailed/pull/826#issuecomment-278724912

**Comment by onox** [comment]
But does the code say: `# Manual of Lycoming engine guarantees engine is fine when oil level >= 2 quarts`? 😄 Do you understand what I mean?

**Comment by gilbertohasnofb** [comment]
> Do you understand what I mean?

Of course, sorry for the retardness today :rofl: 

I will add some comments to the code before we merge this then, I will take a look at this still today

**Comment by onox** [comment]
My concern is just that without explanations, a new developer will have to spend a ton of time reading manuals / specs to verify that some number is correct and not guessed. So then you'll ask yourself the question: this number looks odd, is it because it was (educatedly) guessed or did the dev have/found some manual from somewhere?

**Comment by gilbertohasnofb** [comment]
@onox you are right, of course, I will try to document the best I can and inform where the data come from or where we used educated guesses

**Comment by onox** [comment]
I'm not asking you to write a whole book in the code, just some comments that clarify things a bit 😜 

But having said that, it's good to see you have improved the carburetor! A couple of weeks ago some user mentioned to me he liked that the carburetor is simulated. So thank you for that 😄

**Comment by gilbertohasnofb** [comment]
My pleasure, it has been fun working on this :smile:

**Comment by wlbragg** [comment]
>so that some contributors wouldn't have to worry about accidentally pushing a messy graph/tree of commits

I'm back in the merge business here, it says I can merge it. 
By all means protect it the best you can if it eliminates even one potential mistake.

>But having said that, it's good to see you have improved the carburetor!

I've been testing this extensively on the J3/PA18 and it is really cool.
I got more involved after seeing the supercub video that @kentrosaur posted and I knew I could also use it on the Piper.

**Comment by gilbertohasnofb** [comment]
@onox I just pushed a new commit with hopefully clearer comments on the code, could you please take a look and let me know what you think? If it's all fine, then we can merge this (and possibly that other PR about the bumpspec) and then create a new release

**Comment by gilbertohasnofb** [comment]
@onox @wlbragg @kentrosaur @tigert  thanks a lot for all the testing and feedback, guys.

**Comment by gilbertohasnofb** [comment]
@wlbragg @dg-505 could one of you guys post a little note to the c172p forum page about these modifications to the carburettor system, so that the users pay attention to it and report any bugs? You could basically copy the bullet points I wrote in this PR's description. Thanks a lot!

**Comment by wlbragg** [comment]
Oops, I forgot about that. I took care of it.

**Comment by gilbertohasnofb** [comment]
Great, thanks!
